cmd/mt-whisper-importer-reader/conversion_test.go

```diff
@@ -0,0 +1,1295 @@
+package main
+
+import (
+	"math"
+	"testing"
+
+	"github.com/kisielk/whisper-go/whisper"
+)
+
+func testIncResolution(t *testing.T, inData []whisper.Point, expectedResult map[string][]whisper.Point, method string, inRes, outRes, rawRes uint32) {
+	outData := incResolution(inData, method, inRes, outRes, rawRes)
+
+	if len(expectedResult) != len(outData) {
+		t.Fatalf("Generated data is not as expected:\nExpected:\n%+v\nGot:\n%+v\n", expectedResult, outData)
+	}
+
+	for m, ep := range expectedResult {
+		var p []whisper.Point
+		var ok bool
+		if p, ok = outData[m]; !ok {
+			t.Fatalf("testIncResolution.\nExpected:\n%+v\nGot:\n%+v\n", expectedResult, outData)
+		}
+		if len(p) != len(outData[m]) {
+			t.Fatalf("testIncResolution.\nExpected:\n%+v\nGot:\n%+v\n", expectedResult, outData)
+		}
+		for i := range p {
+			if p[i] != ep[i] {
+				t.Fatalf("Datapoint does not match expected data:\nExpected:\n%+v\nGot:\n%+v\n", expectedResult, outData)
+			}
+		}
+	}
+}
+
+func TestIncResolutionFakeAvgSimple(t *testing.T) {
+	inData := []whisper.Point{
+		{10, 10},
+		{20, 11},
+	}
+
+	expectedResult := map[string][]whisper.Point{
+		"sum": {
+			{5, 50},
+			{10, 50},
+			{15, 55},
+			{20, 55},
+		},
+		"cnt": {
+			{5, 5},
+			{10, 5},
+			{15, 5},
+			{20, 5},
+		},
+	}
+	testIncResolution(t, inData, expectedResult, "fakeavg", 10, 5, 1)
+}
+
+func TestIncResolutionUpToTime(t *testing.T) {
+	inData := []whisper.Point{
+		{10, 10},
+		{20, 11},
+	}
+
+	expectedResult := map[string][]whisper.Point{
+		"sum": {
+			{5, 50},
+		},
+		"cnt": {
+			{5, 5},
+		},
+	}
+	*importUpTo = uint(5)
+	testIncResolution(t, inData, expectedResult, "fakeavg", 10, 5, 1)
+	*importUpTo = math.MaxUint32
+}
+
+func TestIncResolutionFakeAvgNonFactorResolutions(t *testing.T) {
+	inData := []whisper.Point{
+		{10, 10},
+		{20, 11},
+		{30, 12},
+		{40, 13},
+		{50, 14},
+	}
+
+	expectedResult := map[string][]whisper.Point{
+		"sum": {
+			{3, 30},
+			{6, 30},
+			{9, 30},
+			{12, 33},
+			{15, 33},
+			{18, 33},
+			{21, 36},
+			{24, 36},
+			{27, 36},
+			{30, 36},
+			{33, 39},
+			{36, 39},
+			{39, 39},
+			{42, 42},
+			{45, 42},
+			{48, 42},
+		},
+		"cnt": {
+			{3, 3},
+			{6, 3},
+			{9, 3},
+			{12, 3},
+			{15, 3},
+			{18, 3},
+			{21, 3},
+			{24, 3},
+			{27, 3},
+			{30, 3},
+			{33, 3},
+			{36, 3},
+			{39, 3},
+			{42, 3},
+			{45, 3},
+			{48, 3},
+		},
+	}
+
+	testIncResolution(t, inData, expectedResult, "fakeavg", 10, 3, 1)
+}
+
+func TestIncFakeAvgResolutionWithGaps(t *testing.T) {
+	inData := []whisper.Point{
+		{0, 0},
+		{10, 10},
+		{0, 0},
+		{0, 0},
```

Comment:
what exactly is a gap here? as in null data? is this how whisper returns that? a 0 ts and 0 value?

Reply:
yes, because whisper preallocates:
```
$> whisper-create.py whisperfile.wsp 1:6
$> whisper-update.py ./whisperfile.wsp $(date +%s):1
$> whisper-dump.py whisperfile.wsp 
Meta data:
  aggregation method: average
  max retention: 6
  xFilesFactor: 0.5

Archive 0 info:
  offset: 28
  seconds per point: 1
  points: 6
  retention: 6
  size: 72

Archive 0 data:
0: 1504002715,          1
1: 0,          0
2: 0,          0
3: 0,          0
4: 0,          0
5: 0,          0
```